configure.ac

```diff
@@ -337,22 +337,65 @@ if test "x$enable_ubsan" = "xyes"; then
         GF_LDFLAGS="${GF_LDFLAGS} -lubsan"
 fi
 
-# Initialize CFLAGS before usage
-BUILD_TCMALLOC=no
-USE_MEMPOOL=yes
-AC_ARG_WITH([tcmalloc],
-        [AC_HELP_STRING([--without-tcmalloc], [Use gluster based mempool])],
-        [with_tcmalloc="no"], [with_tcmalloc="yes"])
-if test "x$with_tcmalloc" = "xyes"; then
-    AC_CHECK_LIB([tcmalloc], [malloc], [],
-                 [AC_MSG_ERROR([tcmalloc library needs to be present])])
-    BUILD_TCMALLOC=yes
-    GF_CFLAGS="${GF_CFLAGS} -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free"
-    GF_LDFLAGS="-ltcmalloc ${GF_LDFLAGS}"
+dnl Initialize pkg-config magic.
+PKG_PROG_PKG_CONFIG
+
+dnl Select malloc implementation. TCMalloc is preferred since it is
+dnl known to have the best performance on a wide range of benchmarks.
+
+MALLOC_IMPL=tcmalloc
+
+AC_ARG_WITH([malloc],
+            AC_HELP_STRING([--with-malloc=@<:@tcmalloc,jemalloc,system@:>@],
+                             [Use tcmalloc, jemalloc or system default malloc.]))
+case x$with_malloc in
+  x|xtcmalloc)
+    PKG_CHECK_MODULES([TCMALLOC], [libtcmalloc], [use_tcmalloc=yes], [use_tcmalloc=no])
+    if test x$use_tcmalloc = xyes; then
+        AC_DEFINE(USE_TCMALLOC, 1, [Define to 1 if tcmalloc is used.])
+        GF_LDFLAGS="${GF_LDFLAGS} ${TCMALLOC_LIBS}"
+        MALLOC_IMPL=tcmalloc
+    else
+        dnl No tcmalloc detected. If --with-malloc=tcmalloc is specified, this
+        dnl is an error. Otherwise issue warning and fallback to system malloc.
+        if test x$with_malloc = xtcmalloc; then
+            AC_MSG_ERROR([unable to detect tcmalloc, make sure it's installed.])
+        else
+            AC_MSG_WARN([unable to detect tcmalloc, using system malloc instead.])
```

Comment:
I think that we should enforce tcmalloc presence always. If the user doesn't want to install it, an explicit `--with-malloc=system` should be provided to avoid the error.

Reply:
OK.